### PR TITLE
DRAFT Just an inspiration

### PR DESCRIPTION
**Pull request by polquant**
PLS LOOK INTO BUT DO NOT REVIE AND MERGE!!!
Since I do not have time I am sharing the whole branch witch my changes, which one can incorporate and check in your trainings. So far it seems that transformer learns quite slowly, however, it learns ;)

### CODE REVIEW

File: models/feature_extractors/cnn_extrator.py [hunk 1]
```diff
@@ -23,6 +23,9 @@ def __init__(
         )
         # this is a dummy example but in practice this will be longer
 
+        for i, param in enumerate(self.model.parameters()):
+            param.requires_grad = (i >= 297)
```
Comment (polquant) [inline models/feature_extractors/cnn_extrator.py:27]: It seems partial freezing works better for me than freezing scheduler. However we could use number of layers to freeze, instead flag for the whole model part

File: models/model.py [hunk 2]
```diff
@@ -162,24 +180,29 @@ def validation_epoch_end(self, out):
...
                 ]
             )
+            f1 = f1_score(all_targets[nr_head], all_predictions[nr_head],
+                          average='macro', zero_division=0)
             print(head_report)
             if self.run:
-                log_path = "/".join(["metrics/epoch/", head_names[nr_head]])
+                log_path = "/".join(["metrics/epoch/", head_name])
                 self.run[log_path].log(head_report)
+                self.run[f'/metrics/epoch/f1/{head_name}'].log(f1)
```
Comment (polquant) [inline models/model.py:199]: logging to neptune f1 score for all heads separately

File: models/model.py [hunk 3]
```diff
@@ -162,24 +180,29 @@ def validation_epoch_end(self, out):
...
 
         if self.trainer.global_step > 0:
             print("Saving model...")
             torch.save(self.state_dict(), self.model_save_dir)
             self.scheduler.step()
-            if self.freeze_scheduler["freeze_mode"] == "epoch":
+            if (self.freeze_scheduler is not None) and self.freeze_scheduler["freeze_mode"] == "epoch":
```
Comment (polquant) [inline models/model.py:205]: Fix for freeze_scheduler being None

File: models/model.py [hunk 4]
```diff
@@ -206,9 +229,11 @@ def optimizer_step(
     ):
         # set warm-up
         if self.trainer.global_step < self.warmup_steps:
-            lr_scale = min(1.0, float(self.trainer.global_step + 1) / self.warmup_steps)
+            new_lr = min(self.lr,
+                         self.warmup_start_lr +
+                         (self.lr - self.warmup_start_lr) * float(self.trainer.global_step) / self.warmup_steps)
             for pg in optimizer.param_groups:
-                pg["lr"] = lr_scale * self.lr
+                pg["lr"] = new_lr
```
Comment (polquant) [inline models/model.py:236]: warm up phase starting from defined value

File: models/transformers/sign_language_transformer.py [hunk 5]
```diff
@@ -111,19 +111,19 @@ def __init__(
             dropout_rate=dropout_rate,
         )
         self._feedforward_sequential = nn.Sequential(
-            nn.LayerNorm(input_size),
-            nn.Linear(input_size, feedforward_size),
-            nn.ReLU(),
+            nn.Linear(2 * input_size, feedforward_size),
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:114]: In my opinion normalization somehow makes training more difficult

File: models/transformers/sign_language_transformer.py [hunk 6]
```diff
@@ -111,19 +111,19 @@ def __init__(
...
+            nn.ELU(0.1),
             nn.Dropout(dropout_rate),
             nn.Linear(feedforward_size, input_size),
+            nn.ELU(0.1),
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:118]: I prefer ELU insted of RELUS to avoid zeroing of gradient

File: models/transformers/sign_language_transformer.py [hunk 7]
```diff
@@ -133,7 +133,7 @@ def forward(self, input: torch.Tensor):
         attention_output = self._attention_dropout(attention_output)
 
         # Self Attention (Multi-Head Attention) End
-        x = input + attention_output
+        x = torch.concat([input, attention_output], dim=-1)
         feedforward_x = self._feedforward_sequential(x)
```
Comment (polquant) [inline models/transformers/sign_language_transformer.py:137]: give more flexibility to network

File: models/model.py [hunk 8]
```diff
@@ -117,24 +131,28 @@ def __init__(
...
         return predictions
 
     def training_step(self, batch, batch_idx):
         input, target = batch
         targets = target["target"]
         predictions = self(input)
         loss = self.summary_loss(predictions, targets)
-        if self.freeze_scheduler["freeze_mode"] == "step":
+        if (self.freeze_scheduler is not None) and self.freeze_scheduler["freeze_mode"] == "step":
             self.freeze_step()
```
Comment (polquant) [inline models/model.py:156]: Fix for freeze_scheduler being None

File: models/feature_extractors/cnn_extrator.py (hunk 1 above)
Comment (AgaMiko) [reply models/feature_extractors/cnn_extrator.py]: Sure! Sounds like a good improvement!

**Comment by AgaMiko** [comment]
I think it would be nice to merge few of those changes, especially fixes.
